core/state_transition.go

```diff
@@ -344,7 +344,12 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 	if rules.IsLondon {
 		effectiveTip = cmath.BigMin(st.gasTipCap, new(big.Int).Sub(st.gasFeeCap, st.evm.Context.BaseFee))
 	}
-	st.state.AddBalance(st.evm.Context.Coinbase, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveTip))
+	// effectiveTip may be negative during RPC calls such as eth_call and eth_estimateGas.
+	// In these circumstances, we don't want to subtract balance from the coinbase,
+	// which would otherwise result in a potentially negative balance.
+	if effectiveTip.Sign() > 0 {
```

Comment:
The negative tip is an effect of `vm.NoBaseFee` being enabled. That make me think it would be better to go directly after the cause and guard this with `!vm.NoBaseFee` (this is also what is used to skip over some checks in `preCheck()`. The double negative is annoying though.

Reply:
Is there a reason that you want the effective tip to pay out only when it is positive?

Reply:
If a user specifies a gas price for their call in excess of the base fee though, I think it does make sense to give the tip to the coinbase. So if we wanted to gate this on NoBaseFee it might be better to do `if !st.evm.NoBaseFee || effectiveTip.Sign() > 0 {`

Reply:
Hmm right -- okay so I took a look at the PR that introduce `NoBaseFee` (#23027) and there are two motivating use cases:
> * Trying to read smart contract state (we expect gas prices to be irrelevant)
> * Dry running an execution on top of a block (we expect gas prices to be meaningful)

I think the first is satisfied, but the issue that you're trying to address falls under the second. In that case, maybe it is better to accurately simulate the tx if the `feeCap` is non-zero. So in `preCheck()` the `feeCap < baseFee` check could be reworked a bit so that it only is executed under `NoBaseFee = true` when `feeCap` is non-zero.